### PR TITLE
projects: max14906: Add application example for MAX14906

### PR DESCRIPTION
**Pull request by RaduSabau1**
## Pull Request Description

https://github.com/analogdevicesinc/no-OS/pull/1921 needs to be merged before this PR.
This PR contains an application example for MAX14906(both basic and IIO example in one file).
It has been tested on the MAX32666FTHR with the MAX14906PMB.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Comment by buha** [comment]
Worth mentioning, only last commit is up for review here. Maybe put this explicitly in the PR description above.

**Comment by RaduSabau1** [comment]
After #1921 will be merged, I will remove all commits but the last one. I added them here like this if anyone wanted to test the project they won't miss the required files.

File: projects/max14906/Makefile
Comment (RaduSabau1) [reply projects/max14906/Makefile:6]: will add

File: projects/max14906/main.c
Comment (RaduSabau1) [reply projects/max14906/main.c:38]: the iio driver doesn't support buffer since it isn't really needed, so i will remove everything related to this.

File: projects/max14906/main.c
Comment (RaduSabau1) [reply projects/max14906/main.c:103]: Both examples were requested in the same main.c, therefore I used the old project structure.

File: projects/max14906/main.c
Comment (RaduSabau1) [reply projects/max14906/main.c]: Will handle it.

**Comment by RaduSabau1** [comment]
V2

- Removed iio_data_buffer since the driver isn't capable and buffer isn't needed.
- Fixed confusing comments
- Addressed other comments.

File: projects/max14906/main.c
Comment (CiprianRegus) [reply projects/max14906/main.c:98]: In this case, the project will have to live on a separate branch. We can't merge it to master in this form, since it will make further maintenance more error-prone.

File: projects/max14906/main.c
Comment (danmois) [inline projects/max14906/main.c:1]: You should add file description, author, copyright, and license.

Comment (RaduSabau1) [reply projects/max14906/main.c:1]: Adding.

File: projects/max14906/main.c
Comment (RaduSabau1) [reply projects/max14906/main.c:136]: I will use the new project structure after all.

**Comment by RaduSabau1** [comment]
V3

- Switched to new project structure.

**Comment by RaduSabau1** [comment]
V4

- Fixed iio_example errors.

File: projects/max14906/src/examples/dummy/dummy_example.h
Comment (rbolboac) [inline projects/max14906/src/examples/dummy/dummy_example.h:9]: can be moved to .c file seeing how they are not used anywhere else

File: projects/max14906/src/examples/dummy/dummy_example.h
Comment (rbolboac) [inline projects/max14906/src/examples/dummy/dummy_example.h:13]: missing new line at the end of file, this is valid for some other files in this project

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:65]: iio_app_remove and max14906_iio_remove should be called before returning to avoid memory leaks

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:62]: max14906_iio_remove should be called before returning to avoid memory leaks

File: projects/max14906/src/platform/maxim/main.c
Comment (rbolboac) [inline projects/max14906/src/platform/maxim/main.c:32]: missing return

File: projects/max14906/src/examples/dummy/dummy_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/dummy/dummy_example.c:45]: make sure all the remove APIs are called in case of an error to avoid memory leaks, apply to all the returns in this function.

File: projects/max14906/src/examples/dummy/dummy_example.h
Comment (RaduSabau1) [reply projects/max14906/src/examples/dummy/dummy_example.h:9]: Will move.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/iio_example/iio_example.c:65]: Will return.

File: projects/max14906/src/platform/maxim/main.c
Comment (RaduSabau1) [reply projects/max14906/src/platform/maxim/main.c:32]: My mistake.

File: projects/max14906/src/examples/dummy/dummy_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/dummy/dummy_example.c:45]: I am verifying this.

**Comment by RaduSabau1** [comment]
V5

- Fixed memory leaks caused by remove API not being called.
- Addressed other comments.

**Comment by RaduSabau1** [comment]
V16

- Added eof line to files

File: projects/max14906/src/platform/maxim/parameters.h
Comment (CiprianRegus) [inline projects/max14906/src/platform/maxim/parameters.h:4]: wrong email

File: projects/max14906/src/platform/maxim/parameters.c
Comment (CiprianRegus) [inline projects/max14906/src/platform/maxim/parameters.c:4]: here too.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (CiprianRegus) [inline projects/max14906/src/examples/iio_example/iio_example.c:81]: if this fails, you're removing the iio app descriptor, which wasn't even allocated.

File: projects/max14906/src/platform/maxim/parameters.h
Comment (RaduSabau1) [reply projects/max14906/src/platform/maxim/parameters.h:4]: my bad.

**Comment by RaduSabau1** [comment]
V17

- Fixed iio_app_descriptor remove
- Addressed other comments

**Comment by RaduSabau1** [comment]
V17

- Fixed error labels.
- Addressed other comments.

**Comment by RaduSabau1** [comment]
V17

- Fixed error labels.
- Addressed other comments.

**Comment by RaduSabau1** [comment]
V18

- Rearranged commits.

File: projects/max14906/Makefile
Comment (buha) [inline projects/max14906/Makefile:2]: the word `dummy` is not fit, please rename to `basic` or something else

**Comment by RaduSabau1** [comment]
V19

- Dummy Example is now named Basic Example.

File: projects/max14906/Makefile
Comment (amiclaus) [reply projects/max14906/Makefile:2]: @RaduSabau1 this is not resolved.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [reply projects/max14906/src/examples/iio_example/iio_example.c:103]: this is not resolved

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:100]: this should go to error, because if init fails iio_app_remove does not need to be called, only if an error occurs after init is successful.

File: projects/max14906/src/examples/basic/basic_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/basic/basic_example.c:187]: why not use no_os_irq_ctrl_remove instead?... please make sure the _remove APIs are called only if init was successful and an error occurred after, see all cases where it is applicable

File: projects/max14906/src/examples/basic/basic_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/basic/basic_example.c:184]: even if this example is returning successfully, there will be some memory leaks, basically the remove APIs will have to be called because this example allocates some memory which has to be freed before returning.

**Review by rbolboac — COMMENTED** [review]
some remarks from my side related to memory leaks

**Comment by RaduSabau1** [comment]
V21

- Addressed comments about memory leaks.

File: projects/max14906/src/examples/basic/basic_example.c
Comment (buha) [inline projects/max14906/src/examples/basic/basic_example.c:184]: you don't need to remove these before returning 0.

ret is already zero from previous successful function calls, no need to do anything here, remove lines 184-189

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (buha) [inline projects/max14906/src/examples/iio_example/iio_example.c:2]: do we really need such a long path here ?

File: projects/max14906/src/examples/basic/basic_example.c
Comment (buha) [reply projects/max14906/src/examples/basic/basic_example.c:184]: correction, also remove the `return 0`

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/iio_example/iio_example.c:2]: other files had the same. I guessed I should do the same.

**Comment by RaduSabau1** [comment]
V22

- Renamed error labels to be more intuitive.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:81]: max14906_iio_remove should not be called in this case, you can follow the example from this PR: https://github.com/analogdevicesinc/no-OS/pull/1975

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:105]: can be removed

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/iio_example/iio_example.c]: I will and I will apply the same example for the other projects.

**Comment by RaduSabau1** [comment]
V23

- Modified error labels.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/iio_example/iio_example.c:107]: nitpick: rename lablel to iio_max14906_remove

File: projects/max14906/src/platform/maxim/main.c
Comment (rbolboac) [inline projects/max14906/src/platform/maxim/main.c:68]: no_os_uart_remove should be called

Comment (RaduSabau1) [reply projects/max14906/src/platform/maxim/main.c:68]: I will.

File: projects/max14906/src/examples/iio_example/iio_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/iio_example/iio_example.c:107]: will do.

File: projects/max14906/src/examples/basic/basic_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/basic/basic_example.c:91]: will do.

File: projects/max14906/src/examples/basic/basic_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/basic/basic_example.c:103]: will do.

File: projects/max14906/src/examples/basic/basic_example.c
Comment (RaduSabau1) [reply projects/max14906/src/examples/basic/basic_example.c:189]: I will do this and specify in the PR description that it will need to be merged before this PR.

**Comment by RaduSabau1** [comment]
V24

- Calling no_os_uart_remove after basic_example_main.

**Comment by RaduSabau1** [comment]
V24

- Calling no_os_uart_remove after basic_example_main.

**Review by buha — APPROVED** [review]

File: projects/max14906/src/examples/basic/basic_example.c
Comment (rbolboac) [inline projects/max14906/src/examples/basic/basic_example.c:181]: can be removed, it will go anyhow to gpio_irq_unregister_callback

**Review by rbolboac — APPROVED** [review]
One single remark from my side, all looks good otherwise

**Comment by RaduSabau1** [comment]
V25

- Removed unnecessary lines.

**Comment by buha** [comment]
linuxbuilds drivers fails with some timeout, i think it's unrelated to this PR, so I'm merging

**Comment by buha** [comment]
i had to rerun linuxbuilds drivers check (it was failing with timeout) to be able to merge
merging...

**Pull request merged**
